### PR TITLE
Update for resolving tarball location of custom registries to support NPM clients

### PR DESCRIPTION
**Pull request by bkaney**
This is a POC to fix assumptions when trying to download the tarball. See https://chat.fhir.org/#narrow/stream/179239-tooling/topic/NPM.20Registry

### CODE REVIEW

**Comment by cmoesel** [comment]
I like this. Thank you.

Before we can take it as-is, I think we need to determine if this is a valid approach for FHIR registry servers, as they don't document use of `dist.tarball` and instead specify the hard-coded URL pattern [here](https://confluence.hl7.org/display/FHIR/FHIR+Package+Registry+User+Documentation#FHIRPackageRegistryUserDocumentation-Download).

**Comment by cmoesel** [comment]
Oh. And I'm sure it must break some tests too. So I guess we'd need to fix those up as well.

**Comment by bkaney** [comment]
Great! I can work on passing tests / linting / etc.

@cmoesel  - do you think this should only apply with custom registries -- e.g. shall this part get reverted? https://github.com/standardhealth/fhir-package-loader/pull/16/files#diff-51cd61f2399d311debdd2c813260b4a799b0cd6155b8eaa0f1498aebdc5ca281L225-R239 -- then if the FHIR Package spec changes, it could be use for both.

**Comment by bkaney** [comment]
I decided to make that change, need to figure out how `expectDownloadSequence` works, and then should be able to get tests to pass.

**Comment by bkaney** [comment]
I had to modify expectDownloadSequence to indicate when a zipfile was expected (vs when it was not). This is ready to review now @cmoesel

File: test/load.test.ts [hunk 1]
```diff
@@ -591,6 +610,10 @@ describe('#mergeDependency()', () => {
             date: '20200413230227'
           }
         };
+      } else if (uri === 'https://packages.fhir.org/hl7.fhir.r4.core') {
+        return { data: TERM_PKG_RESPONSE };
```
Comment (cmoesel) [inline test/load.test.ts]: The `TERM_PKG_RESPONSE` contents don't match the uri `https://packages.fhir.org/hl7.fhir.r4.core` (i.e., we asked for R4 Core and got back FHIR Terminology). Although the test still works, I'd prefer that it be more consistent with how it works in the real world. Could you change the URI to be the terminology one (`https://packages.fhir.org/hl7.terminology.r4`) or create an `R4_CORE_PKG_RESPONSE` that reflects what R4 core would really look like?

File: test/load.test.ts [hunk 2]
```diff
@@ -591,6 +610,10 @@ describe('#mergeDependency()', () => {
...
+      } else if (uri === 'https://packages2.fhir.org/hl7.fhir.r4.core') {
+        return { data: EXT_PKG_RESPONSE };
```
Comment (cmoesel) [inline test/load.test.ts]: Similar to above, I'd prefer that the response match the query -- so either use uri `https://packages2.fhir.org/packages/hl7.fhir.uv.extensions` or introduce a matching response for R4 Core.

**Review by cmoesel — CHANGES_REQUESTED** [review]
Thanks for the PR, @bkaney! I like the approach!

I just found a couple of nit-picky things in the tests that I'd like updated if possible. Other than that, it looks good and works well. I tested it by exporting `FPL_REGISTRY` as `https://registry.npmjs.org` and then installing `hl7.fhir.us.medmorph@1.0.0` (which Josh Mandel uploaded to NPM the other day). It worked as expected!

**Comment by bkaney** [comment]
@cmoesel - I think this is ready now, LMK if you see anything else.

**Review by cmoesel — APPROVED** [review]
Thank you, @bkaney! It looks great!

We usually require two reviews for PRs, so after someone else from our team reviews it, we'll get it merged!

**Review by jafeltra — APPROVED** [review]
This looks good to me! Thanks!

**Pull request merged**